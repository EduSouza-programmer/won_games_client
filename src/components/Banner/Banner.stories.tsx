import { Story, Meta } from "@storybook/react/types-6-0";
import Banner, { BannerProps } from ".";

export default {
  title: "Banner",
  component: Banner,
  argTypes: {
    ribbon: {
      control: { type: "text" },
    },
    ribbonSize: {
      options: ["normal", "small"],
      control: { type: "select" },
    },
    ribbonColor: {
      options: ["primary", "secondary"],
      control: { type: "select" },
    },
  },
  args: {
    img: "https://images.unsplash.com/photo-1562229125-6d6075419a22?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=967&q=80",
    title: "Defy death",
    subtitle: "<p>Play the new <strong>CrashLands</strong> season",
    buttonLabel: "Buy now",
    buttonLink: "/games/defy-death",
  },
  parameters: {
    layout: "fullscreen",
  },
} as Meta<BannerProps>;

export const Mobile: Story<BannerProps> = (args) => <Banner {...args} />;

Mobile.parameters = {
  viewport: {
    defaultViewport: "mobile1",
  },
};

export const MobileWithRibbon: Story<BannerProps> = (args) => (
  <Banner {...args} />
);

MobileWithRibbon.parameters = {
  viewport: {
    defaultViewport: "mobile1",
  },
};

MobileWithRibbon.args = {
  ribbon: "20% OFF",
  ribbonSize: "small",
  ribbonColor: "primary",
};

export const Desktop: Story<BannerProps> = (args) => (
  <div style={{ width: "104rem", margin: "0 auto" }}>
    <Banner {...args} />
  </div>
);

export const DesktopWithRibbon: Story<BannerProps> = (args) => (
  <div style={{ width: "104rem", margin: "0 auto" }}>
    <Banner {...args} />
  </div>
);

DesktopWithRibbon.args = {
  ribbon: "20% OFF",
  ribbonSize: "normal",
  ribbonColor: "primary",
};
