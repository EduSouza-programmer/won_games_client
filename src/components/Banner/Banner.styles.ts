import styled, { css } from "styled-components";
import media from "styled-media-query";

import * as RibbonStyles from "@/components/Ribbon/Ribbon.styles";

export const Wrapper = styled.div`
  position: relative;

  ${media.lessThan("medium")`

    ${RibbonStyles.Wrapper} {
      right: 0;

      &::before {
         display: none;

        }
      }

    `}

  ${media.greaterThan("medium")`

      box-shadow: 0 0.4rem 0.5rem 0 rgba(0, 0, 0, 0.2);

    `}
`;

const imageBoxModifiers = {
  shimmerEffect: () => css`
    background-image: linear-gradient(
      to right,
      #ffffff 0%,
      #dfdfdf 20%,
      #f6f7f8 40%,
      #f6f7f8 100%
    );
    background-size: 80rem 23rem;
    animation: placeholderShimmer 1s linear infinite forwards;

    @keyframes placeholderShimmer {
      0% {
        background-position: -40rem 0;
      }
      100% {
        background-position: 40rem 0;
      }
    }
    ${media.greaterThan("medium")`
      min-height: 58rem;

    `}
  `,
};

type ImageBox = {
  shimmerEffect: boolean;
};

export const ImageBox = styled.div<ImageBox>`
  ${({ theme, shimmerEffect }) => css`
    width: 100%;
    min-height: 23rem;
    background-color: ${theme.palette.black.main};
    display: flex;
    justify-content: center;

    ${shimmerEffect && imageBoxModifiers.shimmerEffect}

    img {
      width: 100%;
      height: 100%;
      min-height: 23rem;
      object-fit: cover;
    }
  `}
`;

/* type ImageProps = {
  src: string;
};

export const Image = styled.div<ImageProps>`
  ${({ theme, src }) => css`
    width: 100%;
    height: 23rem;
    background-color: ${theme.palette.lightGray.main};
    background-image: url(${src});
    background-position: center center;
    background-size: cover;

    ${media.greaterThan("medium")`
      height: 58rem;
    `}
  `}
`; */

export const Caption = styled.div`
  ${({ theme }) => css`
    width: 100%;
    background-color: rgba(0, 0, 0, 0.7);
    padding: ${theme.spacings.small};
    ${media.greaterThan("medium")`
      border-radius: 0 0 ${theme.border.radius} ${theme.border.radius};
      padding: ${theme.spacings.large};
      position: absolute;
      bottom: 0;
      left: 0;
    `}
  `}
`;

export const Title = styled.h2`
  ${({ theme }) => css`
    font-size: ${theme.typography.fontSizes.large};
    font-weight: ${theme.typography.fontWeightBold};
    padding-bottom: ${theme.spacings.xxsmall};
    color: ${theme.palette.white.main};
    ${media.greaterThan("medium")`
      font-size: ${theme.typography.fontSizes.xxlarge};

    `}
  `}
`;

export const Subtitle = styled.h3`
  ${({ theme }) => css`
    color: ${theme.palette.white.main};
    font-size: ${theme.typography.fontSizes.small};
    font-weight: ${theme.typography.fontWeightMedium};
    padding-bottom: ${theme.spacings.xsmall};
    strong {
      color: ${theme.palette.primary.main};
      font-weight: ${theme.typography.fontWeightBold};
    }

    ${media.greaterThan("medium")`
      font-sizes: ${theme.typography.fontSizes.large};

    `}
  `}
`;
