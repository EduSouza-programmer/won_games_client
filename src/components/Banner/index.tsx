import Ribbon, { RibbonSizes, RibbonColors } from "@/components/Ribbon";
import Button from "@/components/Button";
import * as S from "./Banner.styles";

export type BannerProps = {
  img: string;
  title: string;
  subtitle: string;
  buttonLabel: string;
  buttonLink: string;
  ribbon?: React.ReactNode;
  ribbonSize?: RibbonSizes;
  ribbonColor?: RibbonColors;
};

function Banner({
  img,
  title,
  subtitle,
  buttonLabel,
  buttonLink,
  ribbon,
  ribbonColor = "primary",
  ribbonSize = "normal",
}: BannerProps) {
  return (
    <S.Wrapper>
      <S.Image src={img} role="img" aria-label={title} />
      <S.Caption>
        <S.Title>{title}</S.Title>
        <S.Subtitle dangerouslySetInnerHTML={{ __html: subtitle }} />
        <Button as="a" href={buttonLink} size="large">
          {buttonLabel}
        </Button>
      </S.Caption>
      {!!ribbon && (
        <Ribbon size={ribbonSize} color={ribbonColor}>
          {ribbon}
        </Ribbon>
      )}
    </S.Wrapper>
  );
}

export default Banner;
