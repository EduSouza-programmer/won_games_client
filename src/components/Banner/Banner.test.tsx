import { screen } from "@testing-library/react";
import { renderWithTheme } from "@/utils/tests/helpers";

import Banner from ".";

describe("<Banner />", () => {
  it("should render the heading", () => {
    const { container } = renderWithTheme(<Banner />);

    expect(
      screen.getByRole("heading", { name: /Banner/i })
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
