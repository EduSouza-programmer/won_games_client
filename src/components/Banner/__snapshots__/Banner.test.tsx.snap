// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should render correctly 1`] = `
.c5 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  border-radius: 0.4rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c5:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
}

.c0 {
  position: relative;
  width: 100%;
  height: 100%;
}

.c1 {
  max-width: 104rem;
  max-height: 58rem;
  width: auto;
  height: auto;
  min-height: 23rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #030517;
  height: calc(100% - 40%);
  background-image: linear-gradient( to right,#ffffff 0%,#dfdfdf 20%,#f6f7f8 40%,#f6f7f8 100% );
  background-size: 80rem 58rem;
  -webkit-animation: placeholderShimmer 1s linear infinite forwards;
  animation: placeholderShimmer 1s linear infinite forwards;
}

.c1 img {
  width: 100%;
  height: 100%;
  min-height: 23rem;
  object-fit: cover;
}

.c2 {
  width: 100%;
  height: 100%;
  background-color: rgba(0,0,0,0.7);
  padding: 2.4rem;
}

.c3 {
  font-size: 1.8rem;
  font-weight: 600;
  padding-bottom: 0.8rem;
  color: #FAFAFA;
}

.c4 {
  color: #FAFAFA;
  font-size: 1.4rem;
  font-weight: 400;
  padding-bottom: 1.6rem;
}

.c4 strong {
  color: #F231A5;
  font-weight: 600;
}

@media (max-width:768px) {
  .c0 .sc-bdnxRM {
    right: 0;
  }

  .c0 .sc-bdnxRM::before {
    display: none;
  }
}

@media (min-width:768px) {
  .c0 {
    box-shadow: 0 0.4rem 0.5rem 0 rgba(0,0,0,0.2);
  }
}

@media (min-width:1170px) {
  .c1 {
    height: 58rem;
  }
}

@media (min-width:768px) {
  .c1 {
    border-radius: 0 0 0.4rem 0.4rem;
  }
}

@media (min-width:768px) {
  .c1 {
    height: 100%;
  }
}

@media (min-width:1170px) {
  .c1 {
    height: 58rem;
  }
}

@media (min-width:768px) {
  .c1 img {
    border-radius: 0 0 0.4rem 0.4rem;
  }
}

@media (min-width:768px) {
  .c2 {
    border-radius: 0 0 0.4rem 0.4rem;
    padding: 4.0rem;
    position: absolute;
    bottom: 0;
    left: 0;
    height: auto;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 2.8rem;
  }
}

@media (min-width:768px) {
  .c4 {
    font-sizes: 1.8rem;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="Defy death"
      src="https://source.unsplash.com/user/willianjusten/1042x580"
    />
  </div>
  <div
    class="c2"
  >
    <h2
      class="c3"
    >
      Defy death
    </h2>
    <h3
      class="c4"
    >
      <p>
        Play the new 
        <strong>
          CrashLands
        </strong>
         season
      </p>
    </h3>
    <a
      class="c5"
      href="/games/defy-death"
    >
      <span>
        Buy now
      </span>
    </a>
  </div>
</div>
`;
